### PR TITLE
[CSM] Add `on_punchnode` callback

### PR DESCRIPTION
**Pull request by red-001**
part of #5088

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -3886,17 +3886,20 @@ void Game::handleDigging(GameRunData *runData,
...
 		const ToolCapabilities &playeritem_toolcap, f32 dtime)
 {
+
+	LocalPlayer *player = client->getEnv().getLocalPlayer();
+	ClientMap &map = client->getEnv().getClientMap();
+	MapNode n = client->getEnv().getClientMap().getNodeNoEx(nodepos);
+
 	if (!runData->digging) {
 		infostream << "Started digging" << std::endl;
+		if(client->getScript()->on_punchnode(nodepos, n))
```
Comment (nerzhul) [inline src/game.cpp]: missing space

**Comment by nerzhul** [comment]
Except the space great job !

File: src/game.cpp [hunk 2]
```diff
@@ -3886,17 +3886,20 @@ void Game::handleDigging(GameRunData *runData,
...
 {
+
+	LocalPlayer *player = client->getEnv().getLocalPlayer();
+	ClientMap &map = client->getEnv().getClientMap();
+	MapNode n = client->getEnv().getClientMap().getNodeNoEx(nodepos);
+
 	if (!runData->digging) {
 		infostream << "Started digging" << std::endl;
+		if (client->getScript()->on_punchnode(nodepos, n))
+			return
```
Comment (sfan5) [inline src/game.cpp]: missing ;

Comment (nerzhul) [inline src/game.cpp]: nice i missed it (but not compiler)

**Comment by celeron55** [comment]
I think more care has to be taken in the wordings in doc/client_lua_api.txt overall, but in this PR In particular we shouldn't be talking about "a player" but instead "the local player".

**Comment by PureTryOut** [comment]
I would like to be able to check what item the user is holding when punching a node,. This could be used to change what is going to happen depending on the tool or block the user is holding. Could an argument for that be added?

**Comment by nerzhul** [comment]
@PureTryOut not the purpose of this PR, another PR adding get_wield_item like server side can permit that

**Comment by PureTryOut** [comment]
Fine by me, thanks for your response.

**Comment by paramat** [comment]
@nerzhul @red-001 please link to the 2nd approval.

**Pull request merged**